spec/components/layout_for_admin_spec.rb

```diff
@@ -16,4 +16,25 @@ def component_name
 
     assert_select 'meta[name="robots"][content="noindex,nofollow,noimageindex"]', visible: false
   end
+
+  it "can receive a custom js filename" do
+    allow_any_instance_of(ActionView::Base).to receive(:javascript_include_tag)
```

Comment:
Again just interested to know how this works - why does the JS test have `allow_any_instance_of` and `expect_any_instance_of`, but the CSS test only has `expect_any_instance_of`?

Reply:
Similar to above, the reason we're unable to make an assertion on the page is that it will blow up while trying to render the component if it's unable to find the file based on filename passed into the component. 

Due to that we have to stub out the method that handles the compilation of the assets. It feels like in this situation the thing we really care about is that the correct file is being included.

How allow and expect is that they stub out the method for the class passed in and return "" (by default).
So essentially what happens here is that the layout makes the call to compile `admin.js` and we intercept that call and return an empty string. Blocking it from actually trying to compile and blowing up.

The reason that we need `allow_any_instance_of` in the JS test is due to there being a second method call [here](https://github.com/alphagov/govuk_publishing_components/pull/3993/files#diff-1ebac9542b105f130b13ae04925c85d875d4dc231391387ecb4206b359fc0c07R19) 

So two calls are made:

```
 <%= javascript_include_tag "govuk_publishing_components/vendor/modernizr" %>
```
then
```
 <%= javascript_include_tag "application" %>
```
Without the allow this line

```
 expect_any_instance_of(ActionView::Base).to receive(:javascript_include_tag).with("admin").once
```

blows up because the `javascript_include_tag` method has been stubbed, but only once with expectation that the argument will be "admin".

The output of the test is something similar to

```
expected  javascript_include_tag with the argument of "admin", but it was actually "govuk_publishing_components/vendor/modernizr"
```

The `allow` allows the method to be called with any argument any amount of times which ensures we don't see the error above.  While the expect is checking that the call is made with "admin" once which is the desired out come

Hopefully that makes sense.

Maybe it's clearer if i update the allow to be more explicit and do 

```
allow_any_instance_of(ActionView::Base).to receive(:javascript_include_tag).with("govuk_publishing_components/vendor/modernizr")
 expect_any_instance_of(ActionView::Base).to receive(:javascript_include_tag).with("admin").once
```

So it's clear that the allow is due to the method being called twice, once with a file we're not worried about with our change to the implementation.

Reply:
There are other approaches we could use like dynamically creating an empty `admin.js` file before action before running the test, testing what's in the file, then deleting it after running the test, but that comes with it's own setbacks. Still complex, far slower etc.

But happy to chat it through if you'd like!

I'll go back and update the commit message with some of the above.

Reply:
Really interesting stuff, thanks for taking the time to explain 👍